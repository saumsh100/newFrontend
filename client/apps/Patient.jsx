
import { createBrowserHistory } from 'history';
import { Provider } from 'react-redux';
import ReactDOM from 'react-dom';
import React from 'react';
import moment from 'moment';
import _ from 'lodash';
import * as Immutable from 'immutable';
import 'react-big-calendar/lib/css/react-big-calendar.css';
import socket from '../socket';
import PatientRoutes from '../routes/Patient';
import configure from '../store/availabilitiesStore';

const browserHistory = createBrowserHistory();
const store = configure({ initialState: window.__INITIAL_STATE__, browserHistory });

window.store = store;
window.browserHistory = browserHistory;
window.socket = socket;
window.moment = moment;
window._ = _;
window.Immutable = Immutable;

console.log('Patient App');
console.log('width', window.innerWidth);
console.log('height', window.innerHeight);

document.addEventListener('DOMContentLoaded', () => {
  ReactDOM.render(
    <Provider store={store}>
      <PatientRoutes history={browserHistory} />
    </Provider>,
    document.getElementById('root')
  );
});
