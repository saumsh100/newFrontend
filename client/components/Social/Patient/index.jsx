import React, { Component } from 'react';
import moment from 'moment';
import { Col, Grid, Row, Filters, Calendar } from '../../library';
import colorMap from '../../library/util/colorMap';
import ReviewsCard from './Cards/ReviewsCard';
import styles from './styles.scss';

class Patient extends Component {
  render() {
    const DataBigComment = [{
      icon: 'facebook',
      iconColor: '#ffffff',
      background: '#395998',
      iconAlign: 'flex-end',
      headerLinkName: 'S. Lalala',
      headerLinkSite: 'yelp.ca',
      siteStars: 4,
      siteTitle: 'Lorem Ipsum is simply dummy text of theeMaker including versions of Lorem Ipsum.',
      sitePreview: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheetscontaining Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.',
      createdAt: moment().subtract(10, 'days').fromNow(),
    }, {
      icon: 'bitcoin',
      iconColor: '#ffffff',
      background: '#ffc55b',
      iconAlign: 'center',
      headerLinkName: 'L. Linda',
      headerLinkSite: 'yelp.ca',
      siteStars: 6,
      siteTitle: 'Lorem Ipsum is simply dummy text of theeMaker including versions of Lorem Ipsum.',
      sitePreview: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheetscontaining Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.',
      createdAt: moment().subtract(10, 'days').fromNow(),
    }, {
      icon: 'twitter',
      iconColor: '#ffffff',
      background: '#FF715A',
      iconAlign: 'center',
      headerLinkName: 'N. Blabla',
      headerLinkSite: 'yelp.ca',
      siteStars: 3,
      siteTitle: 'Lorem Ipsum is simply dummy text of theeMaker including versions of Lorem Ipsum.',
      sitePreview: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheetscontaining Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.',
      createdAt: moment().subtract(10, 'days').fromNow(),
    }];

    const filters = [
      {
        title: 'Facebook',
        titleIcon: {
          color: colorMap.facebookBlue,
          icon: 'facebook',
        },
        items: [
          { type: 'checkbox', value: 'ADC Dental Care' },
          { type: 'checkbox', value: 'Carecru Dental Care' },
        ],
      }, {
        title: 'Twitter',
        titleIcon: {
          color: colorMap.lightBlue,
          icon: 'twitter',
        },
        items: [
          { type: 'checkbox', value: 'ADC Dental Care' },
        ],
      }, {
        title: 'Action Filter',
        items: [
          { type: 'checkbox', value: 'Avaiting Response' },
          { type: 'checkbox', value: 'Respond To' },
          { type: 'checkbox', value: 'Dismissed' },
        ],
      },
    ];


    return (
      <Grid className={styles.reviews}>
        <Row>
          <Col className={styles.padding} xs={12} sm={12} md={8} lg={8}>
            <ReviewsCard data={DataBigComment} />
          </Col>
          <Col className={styles.padding} xs={12} sm={4} md={4} lg={4}>
            <Row>
              <Col xs={12} >
                <Filters filters={filters} />
              </Col>
            </Row>
            <Row>
              <Col xs={12} className={styles.calendarContainer}>
                <Calendar className={styles.calendarContainer_calendar} />
              </Col>
            </Row>
          </Col>
        </Row>
      </Grid>
    );
  }
}

export default Patient;
