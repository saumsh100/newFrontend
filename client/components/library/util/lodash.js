
export const isFunction = value => typeof value === 'function';

export const pick = (obj, values) => {
  const newObj = {};
  [...values].forEach((value) => {
    if (obj[value]) {
      newObj[value] = obj[value];
    }
  });
  return newObj;
};

export const omit = (obj, values) => {
  const newObj = { ...obj };
  [...values].forEach((value) => {
    if (newObj[value]) {
      delete newObj[value];
    }
  });
  return newObj;
};

export const isArray = value => Array.isArray(value);

export const clone = (value) => {
  const cloneObj = obj => (typeof obj === 'object' ? { ...obj } : obj);
  return isArray(value) ? [...value] : cloneObj(value);
};

export const forEach = (value, callback) => {
  if (Array.isArray(value) || value instanceof Map) {
    return value.forEach(callback);
  }
  if (typeof value === 'object') {
    const keys = Object.keys(value);
    return keys.forEach(key => callback(value[key]));
  }
  throw new Error('Using forEach is not possible for this variable');
};
