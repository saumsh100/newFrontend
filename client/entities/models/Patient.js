
import createModel from '../createModel';

const PatientSchema = {
  id: null,
  firstName: null,
  lastName: null,
  phoneNumber: null,
  status: null,
  image: null,
  gender: null,
  language: null,
  birthday: null,
};

export default class Patient extends createModel(PatientSchema) {
  /**
   * Add all Patient specific member functions here
   */
  getFullName() {
    return `${this.get('firstName')} ${this.get('lastName')}`;
  }

  getUrlRoot() {
    return `/api/patients/${this.getId()}`;
  }
}
