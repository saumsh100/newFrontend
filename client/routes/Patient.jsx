
import React, { PropTypes } from 'react';
import { Route, Switch } from 'react-router-dom';
import { ConnectedRouter as Router } from 'react-router-redux';
import PatientApp from '../containers/PatientApp';
import FourZeroFour from '../components/FourZeroFour';

const PatientRouter = ({ history }) => {
  return (
    <Router history={history}>
      <div>
        <Route component={PatientApp} />
      </div>
    </Router>
  );
};

PatientRouter.propTypes = {
  history: PropTypes.object.isRequired,
};

export default PatientRouter;
