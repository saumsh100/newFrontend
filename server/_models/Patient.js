
import uniqWith from 'lodash/uniqWith';
import isUndefined from 'lodash/isUndefined';
import isNull from 'lodash/isNull';
import customDataTypes from '../util/customDataTypes';
import { UniqueFieldError } from '../models/createModel/errors';

const STATUS = {
  ACTIVE: 'Active',
  INACTIVE: 'Inactive',
};

export default function (sequelize, DataTypes) {
  const Patient = sequelize.define('Patient', {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },

    accountId: {
      type: DataTypes.UUID,
      allowNull: false,
    },

    pmsId: {
      type: DataTypes.STRING,
    },

    // Used to connect authenticated patientUser to his patient record
    // in the PMS
    patientUserId: {
      type: DataTypes.UUID,
    },

    email: {
      type: DataTypes.STRING,
      validate: {
        isEmail: true,
      },
    },

    firstName: {
      type: DataTypes.STRING,
      allowNull: false,
    },

    lastName: {
      type: DataTypes.STRING,
      allowNull: false,
    },

    middleName: {
      type: DataTypes.STRING,
    },

    phoneNumber: customDataTypes.phoneNumber('phoneNumber', DataTypes),
    homePhoneNumber: customDataTypes.phoneNumber('homePhoneNumber', DataTypes),
    mobilePhoneNumber: customDataTypes.phoneNumber('mobilePhoneNumber', DataTypes),
    workPhoneNumber: customDataTypes.phoneNumber('workPhoneNumber', DataTypes),
    otherPhoneNumber: customDataTypes.phoneNumber('otherPhoneNumber', DataTypes),

    prefContactPhone: {
      // TODO: this should be an enum
      type: DataTypes.STRING,
    },

    /*notes: {
      type: DataTypes.STRING,
    },*/

    gender: {
      // TODO: needs to be an enum
      type: DataTypes.STRING,
    },

    prefName: {
      type: DataTypes.STRING,
    },

    language: {
      // TODO: needs to be an enum
      type: DataTypes.STRING,
    },

    // TODO: needs to be a separate table
    address: {
      // TODO: for now just use nested JSON
      type: DataTypes.JSONB,
      defaultValue: {},
    },

    // TODO: needs to be a separate table
    preferences: {
      type: DataTypes.JSONB,
      defaultValue: {
        morning: true,
        afternoon: true,
        evening: true,
        weekdays: true,
        weekends: true,
        sms: true,
        emailNotifications: true,
        phone: true,
        reminders: true,
        newsletter: true,
        birthdayMessage: true,
      },
    },

    // TODO: what is this for?
    type: {
      type: DataTypes.STRING,
    },

    birthDate: {
      type: DataTypes.DATE,
    },

    // TODO: should be a seperate table
    insurance: {
      type: DataTypes.JSONB,
    },

    isDeleted: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
    },

    isSyncedWithPms: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
    },

    familyId: {
      type: DataTypes.UUID,
    },

    status: {
      type: DataTypes.ENUM,
      values: Object.keys(STATUS).map(key => STATUS[key]),
      defaultValue: STATUS.ACTIVE,
    },
  }, {
    // Model Config
    indexes: [
      {
        unique: true,
        fields: ['accountId', 'email'],
      },
      {
        unique: true,
        fields: ['accountId', 'mobilePhoneNumber'],
      },
    ],
  });

  Patient.associate = ({ Account, Appointment, Chat, SentRecall, Review, SentReview }) => {
    Patient.belongsTo(Account, {
      foreignKey: 'accountId',
      as: 'account',
    });

    Patient.hasMany(Appointment, {
      foreignKey: 'patientId',
      as: 'appointments',
    });

    Patient.hasMany(Chat, {
      foreignKey: 'patientId',
      as: 'chats',
    });

    Patient.hasMany(SentRecall, {
      foreignKey: 'patientId',
      as: 'sentRecalls',
    });

    // This exists because some endpoints refer to 'chats' as 'chat' in the response
    Patient.hasMany(Chat, {
      foreignKey: 'patientId',
      as: 'chat',
    });

    Patient.hasMany(SentReview, {
      foreignKey: 'patientId',
      as: 'sentReviews',
    });

    Patient.hasMany(Review, {
      foreignKey: 'patientId',
      as: 'reviews',
    });
  };

  /**
   *
   * @param model
   * @returns {Promise.<void>}
   */
  Patient.uniqueValidate = async function (model) {
    const { accountId, email, mobilePhoneNumber } = model;
    if (!accountId) {
      throw new Error('model.accountId must exist on the model');
    }

    const noEmail = isUndefined(email) || isNull(email);
    const noMobilePhoneNumber = isUndefined(mobilePhoneNumber) || isNull(mobilePhoneNumber);
    if (noEmail && noMobilePhoneNumber) return;

    // Grab all models that match
    const $or = {};
    if (!noEmail) {
      $or.email = email;
    }

    if (!noMobilePhoneNumber) {
      $or.mobilePhoneNumber = mobilePhoneNumber;
    }

    const p = await Patient.findOne({
      where: {
        accountId,
        $or,
      },
    });

    if (p) {
      throw UniqueFieldError({ tableName: 'Patient' }, 'email or mobilePhoneNumber');
    }
  };

  /**
   *
   * @param dataArray
   * @returns {Promise.<{errors: Array, docs}>}
   */
  Patient.preValidateArray = async function (dataArray) {
    const errors = [];

    const onError = (field, doc) => {
      const error = UniqueFieldError({ tableName: 'Patient' }, field);
      error.patient = doc.dataValues;
      errors.push(error);
    };

    // Build instances of the models
    let docs = dataArray.map(p => Patient.build(p));

    // Now Do ORM Validation
    const validatedDocs = [];
    for (const d of docs) {
      try {
        await d.validate(); // validate against schema
        validatedDocs.push(d);
      } catch (err) {
        err.patient = d.dataValues;
        errors.push(err);
      }
    }

    // Now check uniqueness against each other
    docs = uniqWith(validatedDocs, (a, b) => {
      if (a.accountId && b.accountId && (a.accountId === b.accountId)) {
        if (a.mobilePhoneNumber && b.mobilePhoneNumber && (a.mobilePhoneNumber === b.mobilePhoneNumber)) {
          onError('mobilePhoneNumber', a);
          return true;
        }

        if (a.email && b.email && (a.email === b.email)) {
          onError('email', a);
          return true;
        }
      }
    });

    // Now that they are sanitized, validated, and unique against each other
    const finalDocs = [];
    for (const d of docs) {
      try {
        await Patient.uniqueValidate(d);
        finalDocs.push(d);
      } catch (err) {
        err.patient = d.dataValues;
        errors.push(err);
      }
    }

    docs = finalDocs;
    return { errors, docs };
  };

  Patient.uniqueAgainstEachOther = async (docs) => {
    const errs = [];
    const validDocs = uniqWith(docs, (a, b) => {
      if (a.accountId && b.accountId && (a.accountId === b.accountId)) {
        if (a.dataValues.mobilePhoneNumber && b.dataValues.mobilePhoneNumber
          && (a.dataValues.mobilePhoneNumber === b.dataValues.mobilePhoneNumber)) {
          errs.push(UniqueFieldError({ tableName: 'Patient' }, 'mobilePhoneNumber'));
        }

        if (a.dataValues.email && b.dataValues.email &&
          (a.dataValues.email === b.dataValues.email)) {
          errs.push(UniqueFieldError({ tableName: 'Patient' }, 'email'));
        }
      }
    });

    return { validDocs, errs };
  };


  /**
   *
   * @param dataArray
   * @returns {Promise.<Array.<Model>>}
   */
  Patient.batchSave = async function (dataArray) {
    const { docs, errors } = await Patient.preValidateArray(dataArray);
    const savableCopies = docs.map(d => d.get({ plain: true }));
    const response = await Patient.bulkCreate(savableCopies);
    if (errors.length) {
      const errorsResponse = errors.map((error) => {
        error.model = 'Patient';
        error.errorMessage = 'Patient save error';
        if (error.errors && error.errors[0]) {
          error.errorMessage = error.errors[0].message;
        }
        return error;
      });
      throw { docs: response, errors: errorsResponse };
    }

    return response;
  };

  Patient.allPatients = async function (db, segment) {

  };

  Patient.STATUS = STATUS;

  return Patient;
}
