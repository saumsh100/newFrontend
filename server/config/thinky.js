
const fs = require('fs');
const createThinky = require('thinky');
const globals = require('./globals');

let dbConfig = globals.db;



if (globals.env === 'production') {
  const caCert = fs.readFileSync(`${globals.root}/compose_ca_cert`);
  console.log('globals.caCert');
  console.log(globals.caCert);
  console.log('caCert from file');
  console.log(caCert);
  dbConfig = Object.assign({}, dbConfig, {
    ssl: {
      ca: caCert,
    },
  });
}

const thinky = createThinky(dbConfig);

module.exports = thinky;
