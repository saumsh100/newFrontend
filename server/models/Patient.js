
const thinky = require('../config/thinky');
const createModel = require('./createModel');
const type = thinky.type;

const Patient = createModel('Patient', {
  pmsId: type.string(),
  firstName: type.string().required(),
  middleName: type.string(),
  lastName: type.string().required(),
  phoneNumber: type.string(),
  mobileNumber: type.string(),
  email: type.string().email(),
  notes: type.string(),
  gender: type.string(),
  prefName: type.string(),
  language: type.string(),
  prefContactPhone: type.string(),
  address: type.string(),
  type: type.string(),
  birthDate: type.date(),
  insurance: type.object().allowNull(),

  // TODO: this needs to be modified to support priorities and a standard structure
  appointmentPreference: type.string().enum(['email', 'sms', 'both']).default('both'),
});

// TODO: change to findOne as a general Model function
Patient.defineStatic('findByPhoneNumber', function (phoneNumber) {
  return this.filter({ phoneNumber }).nth(0).run();
});

module.exports = Patient;
