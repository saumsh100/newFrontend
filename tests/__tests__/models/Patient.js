
import { v4 as uuid } from 'uuid';
import { omit } from 'lodash';
import Patient from '../../../server/models/Patient';
import thinky from '../../../server/config/thinky';
import wipeModel from '../../util/wipeModel';

const accountId1 = uuid();
const accountId2 = uuid();

const makeData = (data = {}) => (Object.assign({
  firstName: 'Justin',
  lastName: 'Sharp',
  email: 'justin@carecru.com',
  mobilePhoneNumber: '+18887774444',
  accountId: accountId1,
}, data));

describe('models/Patient', () => {
  beforeEach(async () => {
    await wipeModel(Patient);
  });

  test('should be able to save a Patient without id provided', async () => {
    const data = makeData();
    const patient = await Patient.save(data);
    expect(patient).toMatchObject(data);
  });

  describe('Data Sanitization', () => {
    test('should be able to sanitize Patient mobilePhoneNumber', async () => {
      const data = makeData({ mobilePhoneNumber: '111 222 3333' });
      const patient = await Patient.save(data);

      data.mobilePhoneNumber = '+11112223333';
      expect(patient).toMatchObject(data);
    });
  });


  describe('Data Validation', () => {
    test('should NOT throw Unique Field error', async () => {
      // Save one, then try saving another with same data
      await Patient.save(makeData());
      const data = makeData({ mobilePhoneNumber: '+12222222222', email: 'justin@be.ca' });
      const patient = await Patient.save(data);
      expect(patient.isSaved()).toBe(true);
    });

    test('should NOT throw Unique Field error for diff accountId', async () => {
      // Save one, then try saving another with same data
      await Patient.save(makeData());
      const data = makeData({ accountId: accountId2 });
      const patient = await Patient.save(data);
      expect(patient.isSaved()).toBe(true);
    });

    test('should be able to handle undefined values', async () => {
      const patient = await Patient.save(makeData({ email: undefined }));
      expect(patient.isSaved()).toBe(true);
    });

    test('should throw Unique Field error for 2 docs with same data', async () => {
      // Save one, then try saving another with same data
      await Patient.save(makeData());

      // Couldn't get toThrowError working here...
      try {
        await Patient.save(makeData());
        throw new Error('Did not pass');
      } catch (err) {
        expect(err.message).toBe('Unique Field Validation Error');
      }
    });

    test('should throw Unique Field error for 3 docs with similar data (1 has same number, 1 has same email)', async () => {
      // Save one, then try saving another with same data
      const email = 'justin@be.ca';
      const mobilePhoneNumber = '+18887774444';
      await Patient.save(makeData());
      await Patient.save(makeData({ mobilePhoneNumber: '+12222222222', email }));

      // Couldn't get toThrowError working here...
      try {
        // Same phone number as first, same email as second
        await Patient.save(makeData({ mobilePhoneNumber, email }));
        throw new Error('Did not pass');
      } catch (err) {
        expect(err.message).toBe('Unique Field Validation Error');
      }
    });

    test('should throw Unique Field error for 4 docs with similar data (1 has same number, 1 has same email)', async () => {
      // Save one, then try saving another with same data
      const email = 'justin@be.ca';
      const mobilePhoneNumber = '+12226665555';
      await Patient.save(makeData());
      await Patient.save(makeData({ accountId: accountId2, email }));
      await Patient.save(makeData({ mobilePhoneNumber, email }));

      // Couldn't get toThrowError working here...
      try {
        // Same phone number as first, same email as second
        await Patient.save(makeData({ email }));
        throw new Error('Did not pass');
      } catch (err) {
        expect(err.message).toBe('Unique Field Validation Error');
      }
    });
  });
})













/* OLD TESTS */

const type = thinky.type;

const CARECRU = 'carecru';
const NOT_CARECRU = 'not_carecru';
const PATIENT_TEST_TABLE_NAME = 'PatientTest';

const accountId = 'dc9bd17e-7152-4dbf-9c28-18b1bf5eb30d';
let patientIdGlobal = '';
let patientDocGlobal = {};
let patientDocGlobal2 = false;

let testPatientObject1 = {};
let testPatientObject2 = {};

test.skip('Simple patients write test', () => {
  beforeEach(() => {
    testPatientObject1 = {
      firstName: 'Han',
      lastName: 'Solo',
      mobilePhoneNumber: '1112223344',
      homePhoneNumber: '9998887766',
      accountId,
    };

    testPatientObject2 = {
      firstName: 'Master',
      lastName: 'Yoda',
      mobilePhoneNumber: '5550001122',
      homePhoneNumber: '4440001122',
      accountId,
    };
  });

  afterEach(() => {
    console.log('afterEach: removing test patient: ', patientDocGlobal.id);
    return patientDocGlobal.deleteAll().then(() => {
      const PatientMobilePhoneNumber = Patient.auxModels['mobilePhoneNumber'];
      return PatientMobilePhoneNumber
        .delete()
        .then(() => {
          console.log('All cleaned up.');
        })
        .catch((error) => {
          if (error.name === 'ValidationError') {
            console.log('all good, this still means the test should pass.');
          }
        });
    });
  });

  afterEach(() => {
    if (patientDocGlobal2) {
      return patientDocGlobal2.deleteAll();
    }
  });

  /**
   * - write single patient
   */
  test('Single patient write test and check its aux model', () => {
    console.log('creating testPatientObject1', testPatientObject1);
    return Patient.save(testPatientObject1)
      .then((result) => {
        patientIdGlobal = result.id;
        patientDocGlobal = result;
        console.log('test: result', result);
        expect(result.firstName).toBe(testPatientObject1.firstName);
        expect(result.id).toBe(patientIdGlobal);

        return Patient.auxModels.mobilePhoneNumber
          .get([patientDocGlobal.mobilePhoneNumber, patientDocGlobal.accountId])
          .then((patient1AuxDoc) => {
            expect(patient1AuxDoc.id).toBe(result.id);
            expect(patient1AuxDoc['mobilePhoneNumber.accountId'])
              .toEqual(expect.arrayContaining([result.mobilePhoneNumber, result.accountId]));
          });
      });
  });

  /**
   * - write patient1
   * - write patient2 with the same phone number
   */
  test('Write two patients with same phone number', () => {
    console.log('creating testPatientObject1', testPatientObject1);

    // write first patient and check it and its aux table doc
    return Patient.save(testPatientObject1)
      .then((result) => {
        patientIdGlobal = result.id;
        patientDocGlobal = result;
        expect(result.firstName).toBe(testPatientObject1.firstName);
        expect(result.mobilePhoneNumber).toBe(testPatientObject1.mobilePhoneNumber);

        return Patient.auxModels.mobilePhoneNumber
          .get([patientDocGlobal.mobilePhoneNumber, patientDocGlobal.accountId])
          .then((patient1AuxDoc) => {
            expect(patient1AuxDoc.id).toBe(result.id);

            // write the second patient with the same mobile phone number
            const testPatient2 = omit(testPatientObject1, 'id');

            return Patient.save(testPatient2)
              .then((result2) => {
                throw new Error('Should not have been able to write this patient AGAIN.', testPatient2);
              })
              .catch((err) => {
                expect(err.message).toBe('Unique Field Validation Error');
              });
          });
      });
  });

  /**
   * Testing whether Aux table is updated correctly when a value of the key changes.
   * - write patient1
   * - change phone number of patient 1
   * - change phone number of patient 1 back to original
   */
  test('Create patient, change phone number to diff, change back to the same again.', () => {
    // write patient and check it and its aux table doc
    return Patient.save(testPatientObject1)
      .then((firstSaveResult) => {
        patientIdGlobal = firstSaveResult.id;
        patientDocGlobal = firstSaveResult;

        expect(firstSaveResult.firstName).toBe(testPatientObject1.firstName);
        expect(firstSaveResult.mobilePhoneNumber).toBe(testPatientObject1.mobilePhoneNumber);

        // change mobile phone number of the doc to a new value and persist it
        return Patient.get(firstSaveResult.id)
          .then((doc1) => {
            return doc1.merge({ mobilePhoneNumber: '7780001122' })
              .save()
              .then((secondSaveResult) => {
                expect(secondSaveResult.mobilePhoneNumber).toBe('7780001122');
                expect(secondSaveResult.firstName).toBe(firstSaveResult.firstName);
                expect(secondSaveResult.lastName).toBe(firstSaveResult.lastName);

                // change the phone number back to original and persist it
                return Patient.get(secondSaveResult.id)
                  .then((doc2) => {
                    return doc2.merge({ mobilePhoneNumber: patientDocGlobal.mobilePhoneNumber })
                      .save()
                      .then((thirdSaveResult) => {
                        expect(thirdSaveResult.mobilePhoneNumber).toBe(firstSaveResult.mobilePhoneNumber);
                        expect(thirdSaveResult.firstName).toBe(firstSaveResult.firstName);
                        expect(thirdSaveResult.lastName).toBe(firstSaveResult.lastName);
                      });
                  });
              });
          });
      });
  });

  /**
   * - create two patients with different phone numbres
   * - update one patient with the phone number of another - should now allow that and throw Error
   */
  test('Create two patients, update phone number with existing number', () => {
    return Patient.save([testPatientObject1, testPatientObject2])
      .then((result) => {
        expect(Array.isArray(result)).toBe(true);
        expect(result.length).toBe(2);
        patientDocGlobal = result[0];
        patientDocGlobal2 = result[1];

        // update one of the patients to the phone number of another patient
        return Patient.get(patientDocGlobal.id)
          .then((persistedDoc) => {
            // update patient 1 with the phone number of patient 2
            return persistedDoc.merge({ mobilePhoneNumber: patientDocGlobal2.mobilePhoneNumber })
              .save()
              .then(() => {
                throw Error('Should not happen.');
              })
              .catch((err) => {
                expect(err.message).toBe('Unique Field Validation Error');
              });
          });
      });
  });

  /**
   * - batch create two docs with the same phone number write away
   */
  test.skip('batch create two docs with the same phone number write away', () => {
    Object.assign(testPatientObject2, { mobilePhoneNumber: testPatientObject1.mobilePhoneNumber });
    console.log('modified testPatientObject2', testPatientObject2);
    return Patient.save([testPatientObject1, testPatientObject2])
      .then(() => {
        throw Error('Should not happen');
      })
      .catch((err) => {
        expect(err.message).toBe('Unique Field Validation Error');
      });
  });
});
