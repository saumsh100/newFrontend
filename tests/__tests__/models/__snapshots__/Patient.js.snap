// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models/Patient Data Validation should be able to save a Patient without id provided 1`] = `
Object {
  "accountId": "62954241-3652-4792-bae5-5bfed53d37b7",
  "address": null,
  "birthDate": null,
  "email": "justin@carecru.com",
  "familyId": null,
  "firstName": "Justin",
  "gender": null,
  "homePhoneNumber": null,
  "insurance": null,
  "isDeleted": false,
  "isSyncedWithPms": false,
  "language": null,
  "lastName": "Sharp",
  "middleName": null,
  "mobilePhoneNumber": "+18887774444",
  "otherPhoneNumber": null,
  "patientUserId": null,
  "phoneNumber": null,
  "pmsId": null,
  "prefName": null,
  "prefPhoneNumber": null,
  "preferences": Object {
    "afternoon": true,
    "birthdayMessage": true,
    "emailNotifications": true,
    "evening": true,
    "morning": true,
    "newsletter": true,
    "phone": true,
    "reminders": true,
    "sms": true,
    "weekdays": true,
    "weekends": true,
  },
  "status": "Active",
  "type": null,
  "workPhoneNumber": null,
}
`;
